src/libcore/char.rs

```diff
@@ -419,47 +419,83 @@ impl CharExt for char {
     }
 
     #[inline]
-    fn encode_utf8(self) -> EncodeUtf8 {
-        let code = self as u32;
-        let mut buf = [0; 4];
-        let pos = if code < MAX_ONE_B {
-            buf[3] = code as u8;
-            3
-        } else if code < MAX_TWO_B {
-            buf[2] = (code >> 6 & 0x1F) as u8 | TAG_TWO_B;
-            buf[3] = (code & 0x3F) as u8 | TAG_CONT;
-            2
-        } else if code < MAX_THREE_B {
-            buf[1] = (code >> 12 & 0x0F) as u8 | TAG_THREE_B;
-            buf[2] = (code >>  6 & 0x3F) as u8 | TAG_CONT;
-            buf[3] = (code & 0x3F) as u8 | TAG_CONT;
-            1
-        } else {
-            buf[0] = (code >> 18 & 0x07) as u8 | TAG_FOUR_B;
-            buf[1] = (code >> 12 & 0x3F) as u8 | TAG_CONT;
-            buf[2] = (code >>  6 & 0x3F) as u8 | TAG_CONT;
-            buf[3] = (code & 0x3F) as u8 | TAG_CONT;
-            0
-        };
-        EncodeUtf8 { buf: buf, pos: pos }
+    fn encode_utf8(self, dst: &mut [u8]) -> usize {
+        encode_utf8_raw(self as u32, dst)
     }
 
     #[inline]
-    fn encode_utf16(self) -> EncodeUtf16 {
-        let mut buf = [0; 2];
-        let mut code = self as u32;
-        let pos = if (code & 0xFFFF) == code {
-            // The BMP falls through (assuming non-surrogate, as it should)
-            buf[1] = code as u16;
-            1
-        } else {
-            // Supplementary planes break into surrogates.
-            code -= 0x1_0000;
-            buf[0] = 0xD800 | ((code >> 10) as u16);
-            buf[1] = 0xDC00 | ((code as u16) & 0x3FF);
-            0
-        };
-        EncodeUtf16 { buf: buf, pos: pos }
+    fn encode_utf16(self, dst: &mut [u16]) -> usize {
+        encode_utf16_raw(self as u32, dst)
+    }
+}
+
+/// Encodes a raw u32 value as UTF-8 into the provided byte buffer,
+/// and then returns the number of bytes written.
+///
+/// # Panics
+///
+/// Panics if the buffer is too small.
+#[inline]
+#[unstable(feature = "char_internals",
+           reason = "this function should not be exposed publicly",
+           issue = "0")]
+#[doc(hidden)]
+pub fn encode_utf8_raw(code: u32, dst: &mut [u8]) -> usize {
+    // Marked #[inline] to allow llvm optimizing it away
+    if code < MAX_ONE_B && !dst.is_empty() {
+        dst[0] = code as u8;
+        1
+    } else if code < MAX_TWO_B && dst.len() >= 2 {
+        dst[0] = (code >> 6 & 0x1F) as u8 | TAG_TWO_B;
+        dst[1] = (code & 0x3F) as u8 | TAG_CONT;
+        2
+    } else if code < MAX_THREE_B && dst.len() >= 3  {
+        dst[0] = (code >> 12 & 0x0F) as u8 | TAG_THREE_B;
+        dst[1] = (code >>  6 & 0x3F) as u8 | TAG_CONT;
+        dst[2] = (code & 0x3F) as u8 | TAG_CONT;
+        3
+    } else if dst.len() >= 4 {
+        dst[0] = (code >> 18 & 0x07) as u8 | TAG_FOUR_B;
+        dst[1] = (code >> 12 & 0x3F) as u8 | TAG_CONT;
+        dst[2] = (code >>  6 & 0x3F) as u8 | TAG_CONT;
+        dst[3] = (code & 0x3F) as u8 | TAG_CONT;
```

Comment:
This code unnecessarily does 4 bounds checks, despite you already knowing that destination slice is big enough. You should be using [`get_unchecked_mut`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.get_unchecked_mut) instead of indexing here.

Also in order to duplicate less code between this and `char::utf8_len()`, you could do this:

```
let bytes = char.utf8_len();
if dst.len() < bytes { /* panic */ }
match bytes {
    1 => { /* encode 1 char */ }
    2 => { /* encode 2 chars */ }
    ...
    _ => {}
}
bytes
```